File: backend/app/schemas/session.py
from pydantic import BaseModel, Field

from .query import QueryResponse


class SessionCreate(BaseModel):
    name: str | None = None


class SessionUpdate(BaseModel):
    name: str


class SessionQueryHistoryItem(QueryResponse):
    question: str
    status: str
    error_message: str | None = None


class SessionResponse(BaseModel):
    session_key: str
    name: str | None = None
    messages: list[SessionQueryHistoryItem] = Field(default_factory=list)

    model_config = {
        "from_attributes": True
    }


class SessionListItem(BaseModel):
    session_key: str
    name: str | None = None
    first_message: str | None = None

    model_config = {
        "from_attributes": True
    }


class SessionListResponse(BaseModel):
    items: list[SessionListItem]
    page: int
    page_size: int
    total: int
    total_pages: int

File: frontend/src/types/session.ts
import type { QueryResult } from "./query";

/**
 * Session types.
 * Mirrors `backend/app/schemas/session.py`
 * (`SessionQueryHistoryItem` / `SessionResponse` /
 * `SessionListItem` / `SessionListResponse`)
 * backed by the `sessions` table plus the first user `messages` row.
 */
export type SessionMessage = QueryResult & {
  question: string;
  status: string;
  error_message?: string | null;
};

export type SessionDetail = {
  session_key: string;
  name: string | null;
  messages: SessionMessage[];
};

export type SessionListItem = {
  session_key: string;
  name: string | null;
  first_message: string | null;
};

export type SessionListResponse = {
  items: SessionListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
};

/**
 * Workspace session as used by the UI: the backend list item plus
 * local-only connection state (the backend exposes no connection-status
 * endpoint, so `database` / `type` are only known after connecting
 * in the current page lifetime).
 */
export type Session = SessionListItem & {
  database?: string;
  type?: string;
};

/** Local chat turn state for a session (not persisted by the backend). */
export type Turn = {
  id: string;
  question: string;
  result?: QueryResult;
  error?: string;
};
